File: APCS201906_機器人的路徑.py
n, m = map(int, input().split())
n += 2
m += 2
mat = [[]for i in range(n)]
mat[0] = mat[n-1] = [-1 for i in range(m)]
dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]
for i in range(1, n-1):
    mat[i] = [-1]+[int(x) for x in input().split()]+[-1]
mn = 2000000
for i in range(1, n-1):
    for j in range(1, m-1):
        if mat[i][j] < mn:
            mn = mat[i][j]
            si, sj = i, j
total = 0
while True:
    total += mat[si][sj]
    mat[si][sj] = -1
    mn = 2000000
    ni, nj = si, sj
    for d in range(4):
        i, j = si+dr[d], sj+dc[d]
        if mat[i][j] >= 0 and mat[i][j] < mn:
            mn = mat[i][j]
            ni, nj = i, j
    si, sj = ni, nj
    if mat[si][sj] < 0:
        break

print(total)

File: APCS201906_機器人的路徑.cpp
#include <bits/stdc++.h>
using namespace std;

#define N 110
typedef long long ll;

int mat[N][N];

int main()
{
    freopen("P_7_3_5.in", "r", stdin);
    int n, m;
    cin >> m >> n;
    memset(mat, -1, sizeof(mat));
    int si, sj, mn = INT_MAX;
    for (int i = 1; i <= m; ++i)
        for (int j = 1; j <= n; ++j)
        {
            cin >> mat[i][j];
            if (mat[i][j] < mn)
            {
                mn = mat[i][j];
                si = i, sj = j;
            }
        }

    int di[4] = {-1, 0, 1, 0}, dj[4] = {0, 1, 0, -1};
    int total = 0;
    while (1)
    {
        total += mat[si][sj];
        mat[si][sj] = -1;
        mn = INT_MAX;
        int ni, nj;
        for (int k = 0; k < 4; ++k)
        {
            int i = si + di[k], j = sj + dj[k];
            if (mat[i][j] != -1 && mat[i][j] < mn)
            {
                ni = i, nj = j;
                mn = mat[i][j];
            }
        }
        if (mn == INT_MAX)
            break;
        si = ni, sj = nj;
    }
    cout << total;
    return 0;
}
